docs/algorithms/nautilus/algorithm_en.md

```diff
@@ -0,0 +1,290 @@
+- `--max-tree-size`: Maximum size of generated tree [Default: 1000]
+- `--forksrv`: Enable/disable fork server mode. Default to true. (not recommended to disable)
+
+You can fuzz the calculator like this, for example:
+```
+$ fuzzuf nautilus --out_dir output \
+                  --grammar ./calc_grammar.json \
+                  -- ./test/put_binaries/nautilus/calc @@
+```
+It is working successfully if a screen showing the status of fuzzing in real time appears.
+
+If your usage or grammar file is wrong, an error will show up.
+
+- `Grammar does not exist!`: Grammar file specified by `--grammar` does not exist.
+- `Unknown grammar type ('.json' expected)`: The extension of grammar file is not ".json"
+- `Cannot parse grammar file`: The content of grammar file is wrong. (Check your grammar file as explained in section 2-2.)
+
+## 3. Algorithm
+Generally, grammar based fuzzers generates AST (abstract syntax tree) according to a specific grammar and mutates a part of the AST to create testcases. Nautilus internally only uses the tree representation and mutates the tree.
+In this section, we explain the design of how Nautilus generates and mutates the testcases.
+
+### 3-1. Generation
+Since a non-terminal may have multiple rules, we need an algorithm to decide which rule to pick up. Nautilus uses uniform generation algorithm.
+Let's consider the following grammar:
+```
+<PROG> := <STMT>
+<PROG> := <STMT>; <PROG>
+<STMT> := return 1
+<STMT> := <VAR> = <EXPR>
+<VAR>  := a
+<EXPR> := <NUMBER>
+<EXPR> := <EXPR> + <EXPR>
+<NUMBER> := 1
+<NUMBER> := 2
+```
+For example, `<STMT>` has 2 rules: `return 1` or `<VAR> = <EXPR>`. If we choose a rule for each non-terminal by naive randomness, `return 1` is chosen with 50% probability. On the other hand, if we choose `<VAR> = <EXPR>`, `<EXPR>` has another 2 rules: `<NUMBER>` and `<EXPR> + <EXPR>`. The probability that one of them is chosen is 25% from `<STMT>`.
+The deeper part of the tree is selected with less probability with a naive randomness like this, which results in generating similar testcases. Naitlus, on the other hand, uses an algorithm by McKenzie[^2] so that it can select every rule in the grammar with the same probability.
+
+### 3-2. Minimization
+Nautilus attempts to create a smaller testcase that triggers the same new coverage after it found an interesting input. Minimized inputs can make the execution time shorter and the number of set of potential mutations smaller. Nautilus uses two approaches to minimize the testcase that found new paths.
+
+#### 3-2-a. Subtree Minimization
+**Subtree Minimization** is a process to make the subtree of AST as short as possible.
+We generate the smallest possible subtree for each non-terminal. Then, we replace the subtree of each node sequentially and check if we get the same coverage as that of the original tree. If we get the same coverage, the replace tree is taken and otherwise the change is discarded.
```

Comment:
```suggestion
We generate the smallest possible subtree for each non-terminal. Then, we replace the subtree of each node sequentially and check if we get the same coverage as that of the original tree. If we get the same coverage, the replaced tree is taken and otherwise, the change is discarded.
```